### PR TITLE
feat: use same server-info file for all map modes

### LINKED ISSUE #1836: Numaflow single-server info for map

### PR DESCRIPTION
**Pull request by kohlisid**
Fixes #1836

### CODE REVIEW

File: pkg/udf/map_udf.go [hunk 1]
```diff
@@ -193,14 +189,11 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
...
 		opts = append(opts, forward.WithUDFBatchMap(batchMapHandler))
-	} else {
-		// Default is to enable unary map mode
 
-		// Wait for server info to be ready
-		serverInfo, err := sdkserverinfo.SDKServerInfo(sdkserverinfo.WithServerInfoFilePath(sdkclient.MapServerInfoFile))
-		if err != nil {
-			return err
-		}
+	} else if !ok || mapMode == string(sdkserverinfo.UnaryMap) {
```
Comment (whynowy) [inline pkg/udf/map_udf.go:193]: `else {`?

Log the mode.

File: pkg/udf/map_udf.go [hunk 2]
```diff
@@ -167,14 +168,9 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
...
-	} else if enableBatchMapUdf {
+	} else if ok && mapMode == string(sdkserverinfo.BatchMap) {
 		// if Batch Map mode is enabled create the client and handler for that accordingly
-
-		// Wait for server info to be ready
-		serverInfo, err := sdkserverinfo.SDKServerInfo(sdkserverinfo.WithServerInfoFilePath(sdkclient.BatchMapServerInfoFile))
-		if err != nil {
-			return err
-		}
+		enableBatchMapUdf = true
```
Comment (whynowy) [inline pkg/udf/map_udf.go:173]: Log the mode.

File: pkg/udf/map_udf.go [hunk 3]
```diff
@@ -132,22 +132,23 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
...
+	mapMode, ok := serverInfo.Metadata[sdkserverinfo.MapModeMetadata]
+
+	if ok && mapMode == string(sdkserverinfo.StreamMap) {
 		// Map Stream mode
-		// Wait for server info to be ready
-		serverInfo, err := sdkserverinfo.SDKServerInfo(sdkserverinfo.WithServerInfoFilePath(sdkclient.MapStreamServerInfoFile))
-		if err != nil {
-			return err
-		}
+		enableMapUdfStream = true
```
Comment (whynowy) [inline pkg/udf/map_udf.go:151]: Log the mode.

File: pkg/udf/map_udf.go [hunk 4]
```diff
@@ -132,22 +132,24 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
...
+	if err != nil {
+		return err
 	}
 
-	if enableMapUdfStream {
+	// Read the server info file to read which map mode is enabled
+	// Based on the value set, we will create the corresponding handler and clients
+	mapMode, ok := serverInfo.Metadata[sdkserverinfo.MapModeMetadata]
+
+	if ok && mapMode == string(sdkserverinfo.StreamMap) {
```
Comment (whynowy) [inline pkg/udf/map_udf.go:149]: `if ok && sdkserverinfo.MapMode(mapMode) = sdkserverinfo.StreamMap` {

**Review by ayildirim21 — APPROVED** [review]

File: pkg/sdkclient/serverinfo/serverinfo.go [hunk 5]
```diff
@@ -31,6 +31,18 @@ import (
 	"github.com/numaproj/numaflow"
 )
 
+// MapModeMetadata field is used to indicate which map mode is enabled
+// If none is set, we consider unary map as default
+const MapModeMetadata = "MAP_MODE"
```
Comment (whynowy) [inline pkg/sdkclient/serverinfo/serverinfo.go:36]: Get all the `metadata` keys together in this file? e.g. `MULTIPROC`.

Comment (kohlisid) [reply pkg/sdkclient/serverinfo/serverinfo.go]: Makes sense!

**Review by whynowy — APPROVED** [review]

**Review by vigith — APPROVED** [review]

**Pull request merged**